refactor(forEach): change to for-of with iterable

PatrickJS opened the pull request:
Tests are passing. I also used `git mv` to preserve the git history. I changed references with the name array to iterable
closes #909

vicb commented:
@gdi2290 the Travis error comes from the benchmarks. infinite_scroll uses `ForEach/For` and needs to be updated

gdi2290 commented:
@vicb thanks I couldn't make out what the Travis error was. Give me a second

gdi2290 commented:
@vicb Dart doesn't have a for-of so how do we change the interface for Dart users to use for-in. Also where is the bug for the es6 version (for...in)?

gdi2290 commented:
:/ I'm doing something wrong with my Dart version
```dart
void iterateListLike(iter, fn(item)) {
  if (iter is List) {
    for (var item = 0; item < iter.length; ++item) {
      fn(iter[item]);
    }
  } else {
    for (var item in iter) {
      fn(item);
    }
  }
}
```

gdi2290 commented:
@mhevery I rebased and fixed the merge conflicts

gdi2290 commented:
is there a better way to rebase on github? I always force push up changes after rebase but that also blows up any commits that were made in the files previously :/

gdi2290 commented on modules/angular2/test/directives/for_spec.js:
```diff
@@ -124,8 +124,8 @@ export function main() {
 
-    it('should iterate over an array of objects', () => {
-      compileWithTemplate('<ul><li template="foreach #item in items">{{item["name"]}};</li></ul>').then((pv) => {
+    it('should iterate over an array of objects', inject([AsyncTestCompleter], (async) => {
```
I also fixed the build in this commit
[foreach_spec.js#L127])https://github.com/angular/angular/blob/afda43dc0281e3671d6df3218749d5050d47f188/modules/angular2/test/directives/foreach_spec.js#L127)

mhevery commented:
In general I like this commit. You did make the change detection slower however. Those parts need to be reverted, but the rest can go. Please ping this thread once ready.

gdi2290 commented:
The commit was squash without other changes. I also preserved the git history during renaming. There needs to be Iterator facade next and then a perf PR to avoid `polymorphic` code for when it's an iterable

caitp commented:
there are only a handful of iterables most people are going to use, it's not likely to result in IC misses if the path is only taken when the object is known statically to be iterable

gdi2290 commented:
gg I have to update the TodoMVC example to use `For`

PatrickJS committed:
refactor(forEach): change to for-of with iterable

rename: foreach -> for
rename: array -> iterable
update: DartParseTreeWriter
update: naive_infinite_scroll
update: todo
fix: tests in foreach_spec

gdi2290 commented:
@mhevery PR is all green now with correct rebase and renames tracked by git

gdi2290 commented:
@mhevery I rebase'd Master again so everything is green. If you have any suggestions or changes feel free to ping me again about it

mhevery commented:
b61b8d60b7cfd10786ccd9ba4071738d4f2bf15a
